bpf/pca.h

```diff
@@ -0,0 +1,106 @@
+#ifndef __PCA_H__
+#define __PCA_H__
+
+#include "utils.h"
+#include <string.h>
+
+static int attach_packet_payload(void *data, void *data_end, struct __sk_buff *skb, u64 current_time ){
+        payload_meta meta;
+        u64 flags = BPF_F_CURRENT_CPU;
+        // Enable the flag to add packet header
+        // Packet payload follows immediately after the meta struct
+        u32 packetSize = (__u32)(data_end-data);
+        // For packets which are allocated non-linearly struct __sk_buff does not necessarily 
+        // has all data lined up in memory but instead can be part of scatter gather lists. 
+        // This command pulls data from the buffer but incurs data copying penalty.
+        if (packetSize < skb->len){
+            packetSize = skb->len;
+            if (bpf_skb_pull_data(skb, skb->len)){
+                return TC_ACT_UNSPEC;
+            };
+        }
+        // Set flag's upper 32 bits with the size of the paylaod and the bpf_perf_event_output will 
+        // attach the specified amount of bytes from packet to the perf event
+        // https://github.com/xdp-project/xdp-tutorial/tree/9b25f0a039179aca1f66cba5492744d9f09662c1/tracing04-xdp-tcpdump       
+        flags |= (__u64)packetSize << 32;
+
+        meta.if_index = skb->ifindex;
+        meta.pkt_len = packetSize;
+        meta.timestamp = current_time;
+        if (bpf_perf_event_output(skb, &packet_record, flags, &meta, sizeof(meta))){
+            return TC_ACT_OK;
+        }
+        return TC_ACT_OK;
+}
+
+static inline bool validate_pca_filter(u8 ipproto, void *ipheaderend, void *data_end){
+    //Only export packets with protocol set by ENV var PCA_FILTER
+    u16 sourcePort, destPort;
+    if (ipproto != pca_proto) {
+       return false;	
+    }
+
+    if (ipproto == IPPROTO_TCP){
+        struct tcphdr *tcp_header = ipheaderend;
+        if ((void *)tcp_header + sizeof(*tcp_header) > data_end) {
+            return false;
+        }
+        sourcePort = tcp_header->source;
+        destPort = tcp_header->dest;
+    }
+    else{
+        struct udphdr *udp_header = ipheaderend;
+        if ((void *)udp_header + sizeof(*udp_header) > data_end) {
+            return false;	
+        }
+        sourcePort = udp_header->source;
+        destPort = udp_header->dest;
+    }
+    if (sourcePort == bpf_htons(pca_port) || destPort == bpf_htons(pca_port)){
+        return true;
+    }
+    return false;
+}
+
+static inline int export_packet_payload (struct __sk_buff *skb) {
+    void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    
+    struct ethhdr *eth  = data;
+    struct iphdr  *ip;
+
+    // Record the current time.
+    __u64 current_time = bpf_ktime_get_ns();
+    
+    if ((void *)eth + sizeof(*eth) > data_end) {
+       return TC_ACT_UNSPEC;
+    }
+
+    // Only IPv4 and IPv6 packets captured
+    if (eth->h_proto != bpf_htons(ETH_P_IP) && (eth->h_proto != bpf_htons(ETH_P_IPV6))) {
```

Comment:
might be more efficient instruction wise to do
```c
u16 ethType = bpf_ntohs(eth->h_proto);
if (ethType != ETH_P_IP && ethType != ETH_P_IPV6) {
}
```

Reply:
Updated.